Fix Rollback/Restore by sorting record list from database.

cartman-2000 opened the pull request:
Fixes #47 
This fixes the rollback and restore for the h2 and mysql backends by sorting the records in the query, rollback/restore is order dependant on the records created time.  Restore must have an ascending order from oldest to newest, and a rollback requires a descending order, from newest to oldest. The only instance that a rollback/restore might not work properly with this is if there's multiple block records at the same location in the same second.

This also changes the order in the lookup tool to show the newest records first, as most of the time someone would be looking for the newest records in a lookup.

Note: This however doesn't include the ordering fix for the Mongo backend, as I have no setup of this database engine to test with.

viveleroi commented on src/main/java/com/helion3/prism/api/query/QuerySession.java:
```diff
@@ -140,4 +142,20 @@ public void setRadius(int radius) {
+     * Sets the Applier mode used in the Applier command.
+     * @param applier
+     */
+    public void setApplier(ApplierMode applier) {
```
One of my goals with 3.0 is to abstract this kind of data out, because I want this flexibility to be available without tying the query system to Prism internals (appliers). I feel like a new flag would be a good choice here, we could potentially allow users to customize sorts, and flags are a custom structure we already support, but at the very least I think all we need here is an enum for sort direction. We could possibly restrict what kinds of sorts we allow, with something like `Sort.NEWEST_FIRST`.

cartman-2000 replied:
I could update the constructor QuerySession(CommandSource commandSource) in this class to include the sort parameter, so you would have to define the sorting to be used when creating an instance of QuerySession.

viveleroi replied:
I think at least half of the time it won't be necessary, and an API use may want to change it dynamically so it should remain a getter/setter (unless it's a flag, then it would use that system).

viveleroi commented on src/main/java/com/helion3/prism/storage/h2/H2SQLQuery.java:
```diff
@@ -39,18 +40,24 @@ public H2SQLQuery(String query) {
      */
     public static SQLQuery from(QuerySession session) {
-        Builder query = SQLQuery.builder().select().from(tablePrefix + "records");
+        Builder query = SQLQuery.builder().select().from(tablePrefix + "records AS a");
```
Please use something a bit more descriptive as an alias, "a" and "b" aren't going to work. For one-letter aliases like these, I typically use the first letter of the table, so `r` and `e` in this case.

viveleroi commented on src/main/java/com/helion3/prism/storage/h2/H2SQLQuery.java:
```diff
@@ -39,18 +40,24 @@ public H2SQLQuery(String query) {
 
-        if (session.hasFlag(Flag.NO_GROUP)) {
-            query.order("created", "y", "x", "z");
+        if (session.getApplier().isPresent() && session.getApplier().get() == ApplierMode.RESTORE) {
```
Once we have a flag or a`Sort` enum they shouldn't be optional. If the user doesn't provide one (null shouldn't be acceptable) we should return a default.

viveleroi commented on src/main/java/com/helion3/prism/util/BlockUtil.java:
```diff
@@ -112,7 +112,6 @@ public static boolean rejectPlaceCombination(BlockType a, BlockType b) {
 
             // It's fire (which didn't burn anything)
             (a.equals(BlockTypes.FIRE) && b.equals(BlockTypes.AIR)) ||
-            (a.equals(BlockTypes.AIR) && b.equals(BlockTypes.FIRE)) ||
```
Can you clarify why you removed this and added the check to the eventutil?

cartman-2000 replied:
It's to allow the logging of fire placed by the player(to track who set a fire.), but fire placed by environment still won't be logged like before.

cartman-2000 commented:
It's been updated with the recommendations, I also added the flag oldest(with an alias of old), that flips the lookup query to oldest first.

I also updated the comment for the fire tracking checks.

cartman-2000 commented:
I also fixed a bug on the block parameter, as the block id's are saved to the database with the underscores changed to spaces, the same has to be done for the parameter to return any blocks if the block id has underscores in it.

viveleroi commented on src/main/java/com/helion3/prism/api/flags/FlagOldest.java:
```diff
@@ -0,0 +1,59 @@
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+
+public class FlagOldest extends SimpleFlagHandler {
```
My intention was that we'd have a flag for sorting in general, so this would be `FlagSort` and would parse the user input to determine a direction, "asc" or "desc" which we can map to the enum you made. In the future sometime, we can expand on this and allow them to specify a *field* too, so someone can truly customize how the results are sorted.

cartman-2000 replied:
I had thought about doing a parameter for sorting order, but for simplicity sake I went for this as having an order by newest in the command would be redundant at the current moment in time, with the default sort order to be newest first.

Although a parameter sort order would be more useful if there was a config option added to the plugin config that allowed the default sort order to be set.

viveleroi replied:
It would be best to get that sort flag added in now, even if it's extremely limited, so that we never have to break functionality or change syntax on people. Rather than introducing people to `-oldest` we can do `-sort=desc` and in the future we can add an optional field parameter. That way they don't have to re-learn the syntax.

viveleroi commented on src/main/java/com/helion3/prism/api/query/QuerySession.java:
```diff
@@ -45,24 +45,31 @@
      */
     public QuerySession() {
         this.commandSource = null;
+        this.sort = Sort.NEWEST_FIRST;
```
Since it's default value we don't change between constructors, you can define this inline on line 48. 

     protected Sort sort = Sort.NEWEST_FIRST;

viveleroi replied:
This change also means the added comments to these constructors aren't needed.

viveleroi commented on src/main/java/com/helion3/prism/api/query/Sort.java:
```diff
@@ -0,0 +1,43 @@
+    NEWEST_FIRST(-1, "DESC"),
+    OLDEST_FIRST(1, "ASC");
+
+    private int sortVal;
```
Please add one new line between all of these - the fields, constructor, and functions? Thanks.

viveleroi commented on src/main/java/com/helion3/prism/commands/ApplierCommand.java:
```diff
@@ -45,11 +46,7 @@
-    }
-
-    public static CommandSpec getCommand(ApplierMode mode) {
+    public static CommandSpec getCommand(Sort sort) {
```
Why is `sort` a parameter here when that value should be provided by the user via a flag?

cartman-2000 replied:
While sorting can be handled by user input in the normal lookup commands, it needs to forced here(To ignore a user order flag.) for proper rollback/restore, as these are order sensitive to created time. Without the forced sorting a roll back wouldn't work properly if the user used the flag for oldest first sorting.

viveleroi commented on src/main/java/com/helion3/prism/storage/mongodb/MongoRecords.java:
```diff
@@ -253,15 +248,9 @@ else if (field.getMatchRule().equals(MatchRule.BETWEEN)) {
+       // Sorting. Newest first for rollback and oldest first for restore.
        Document sortFields = new Document();
-       sortFields.put(DataQueries.Created.toString(), sortDir);
-       sortFields.put(DataQueries.Y.toString(), 1);
```
The reason I had a Y sort in here stems from Prism 1, it's possible that multiple records from a bulk change have the same timestamp, which means that those records may be out of order. If this is sand, we should restore from the bottom up.

I don't think it's as critical for X/Z sorts though.

cartman-2000 replied:
In my testing, there hasn't been an issue with rolling back falling sand when it is sorted by the created column only. 

With BlockChangeFlag.NONE on the BlockSnapshot restore operations, the neighboring blocks won't be updated/triggered in a rollback/restore, which would be good for falling type blocks as physics won't be performed on them when they're handled, so they can be handled out of order on the y axis without falling. The most important form of sorting for rollbacks/restores is by sorting by created time, sorting by location isn't as important, and there can be some potential issues with multi column sorting.

cartman-2000 commented:
Uploaded changes, Changed formatting of the Sort class, and used the recommendation in the QuerySession class.

cartman-2000 commented:
Uploaded changes, changed the old flag to the order flag, with how the flag handling code works, using an empty flag -ord or -order will have the same effect as the old flag did(.acceptsValue() isn't called when there's not value for the flag.). The accepted values for the flag are: new, newest, desc, for newest first sorting, and: old, oldest, asc, for oldest first sorting.
